### DISCUSSION TITLE
Required classes with optional properties

### DISCUSSION THREAD

**Discussion by ctigrisht** [dotnet/csharplang#8175] [Language Ideas]
Hey! So on C# discord we were discussing about the possibility of having a `required` or `strong` class where all member properties are automatically required and optional properties need to be marked with `optional`.

Usage would look like so:
```cs
public strong class MyNewRequiredClass {
  public string SomeRequiredString {get; set;} // this one would be required
  public optional string? SomeOptionalString {get;set;} // this one is optional on init
  public string SomeDefaultValuedString {get;set;} = "Jacob is not my name"; // this one is optional as it has a default value
}
```
Alternatively as pointer out by iam3yal we could have something like this without introducing new keywords:
```cs
public required class MyNewRequiredClass {
  public string SomeRequiredString {get; set;} // this one would be required
  public string? SomeOptionalString {get;set;} = _; // this one is optional on init
  public string SomeDefaultValuedString {get;set;} = "Jacob is not my name"; // this one is optional as it has a default value
}
```

It would greatly help us in our workflow as 90+% of properties we use are marked as `required` anyways. Would love to see it in the language.

**Comment by tacosontitan** [comment]
I'm all for explicit specification and not introducing new keywords unless we have to. With that in mind, if such a feature was implemented, can we treat marking a property as nullable as the explicit "optional" instead of a new keyword? What's the impact of going that route in contrast with introducing a new keyword?

I'm not sure I'm for or against this right now.

**Reply by ctigrisht** [reply]
Well I think that marking as optional explicitly would be better, as sometimes you may want to point attention that the required nullable property being forgotten should actually not be null in the specific case.

**Reply by HaloFour** [reply]
It's also an explicit design choice that nullable and optional are two completely different concerns.  You can have a required nullable property and an optional non-nullable property.

**Comment by iam3yal** [comment]
It might be possible to use a discard `_` as opposed to yet another keyword? and if `required` is the default then I'd stick with that and put it on the class declaration so something like this:
```csharp
public required class MyNewRequiredClass {
  public string SomeRequiredString {get; set;} // this one would be required
  public string? SomeOptionalString {get;set;} = _ // this one is optional on init
  public string SomeDefaultValuedString {get;set;} = "Jacob is not my name"; // this one is optional as it has a default value
}
```

**Reply by ctigrisht** [reply]
yeah I think this is good, upvote

**Comment by mgelei** [comment]
I'm wondering how prevalent is that "90+% of properties [being] marked as required", but if I look at it purely from a language perspective, I'd be cautious about magic keywords that flip the behaviour of language constructs.

If I look at a property now I'm able to tell how it behaves. By introducing class-wide `strong` or similar, I'd have to look at another place to verify if I'll get the expected behavior, or I'd have to keep that piece of extra information in mind.
